cf/data/data.py

```diff
@@ -3831,304 +3831,70 @@ def _binary_operation(self, other, method):
+                result = np.ma.masked_invalid(dx0, copy=False)
+                np.seterr(**_seterr)
+            else:
+                # Raise the floating point error exception
+                raise FloatingPointError(error)
+        except TypeError as error:
+            if inplace:
+                raise TypeError(
+                    "Incompatible result data-type ({0!r}) for "
+                    "in-place {1!r} arithmetic".format(
+                        np.result_type(dx0.dtype, dx1.dtype).name,
+                        dx0.dtype.name,
                     )
-                else:
-                    raise TypeError(error)
-            # --- End: try
-
-            if array0 is NotImplemented:
-                array0 = np.zeros(partition.shape, dtype=bool)
-            elif not array0.ndim and not isinstance(array0, np.ndarray):
-                array0 = np.asanyarray(array0)
-
-            if not inplace:
-                p_datatype = array0.dtype
-                if new_dtype != p_datatype:
-                    new_dtype = np.result_type(p_datatype, new_dtype)
-
-            partition.subarray = array0
-            partition.Units = new_Units
-            partition.axes = new_axes
-            partition.flip = new_flip
-            partition.part = []
-
-            if broadcasting:
-                partition.location = dummy_location
-                partition.shape = list(array0.shape)
-
-            partition._original = None
-            partition._write_to_disk = False
-            partition.close(units=new_Units)
-
-            if not inplace:
-                partition_s.close()
-        # --- End: for
-
-        # Reset numpy.seterr
-        np.seterr(**original_numpy_seterr)
-
-        source = result.source(None)
-        if source is not None and source.get_compression_type():
-            result._del_Array(None)
-
-        if not inplace:
-            result._Units = new_Units
-            result.dtype = new_dtype
-            result._flip(new_flip)
-
-            if broadcasting:
-                result.partitions.set_location_map(result._axes)
-
-            if method_type in ("_eq", "_ne", "_lt", "_le", "_gt", "_ge"):
-                result.override_units(Units(), inplace=True)
-
-            return result
-        else:
-            # Update the metadata for the new master array in place
-            data0._shape = new_shape
-            data0._ndim = new_ndim
-            data0._size = new_size
-            data0._axes = new_axes
-            data0._flip(new_flip)
-            data0._Units = new_Units
-            data0.dtype = new_dtype
-
-            if broadcasting:
-                data0.partitions.set_location_map(new_axes)
-
-            self.__dict__ = data0.__dict__
+                )
+            else:
+                raise TypeError(error)
 
+        if inplace:  # in-place so concerns original self
+            self._set_dask(result, reset_mask_hardness=False)
```

Comment:
```suggestion
            self._set_dask(result)
```

Reply:
Oh, I thought not resetting the mask hardness would be appropriate here? Happy to make this change (in the two appropriate locations raised in these comments), but just wondering why this is preferred?

Reply:
The practical reason is that `reset_mask_hardness` is no longer a keyword of `_set_dask`! See #399 for more details ...

Reply:
That's a very good reason! Sorry, I do remember this - again it is something that I neglected to update given I started work on this branch months ago... I'll commit those now, thanks.